### DISCUSSION TITLE
What's the most efficient way to write any/all calls with nested type-checked path expressions?

### DISCUSSION THREAD

**Discussion by isidentical** [geldata/gel#2051] [Q&A]
I am trying to produce the most efficient `all()`/`any()` expression output that I can think of, but I'm kind of stuck on nested operations. I am using this [schema](https://github.com/reizio/reiz.io/blob/master/static/Python-reiz.edgeql)

For simple expressions like filtering a set by the type, I use this form;
```
SELECT
    ast::FunctionDef
FILTER
    all(.body IS ast::Expr)
LIMIT 5;
```

but if I try to use this form in a compound way (like type checking etc.)
```
SELECT
    ast::FunctionDef
FILTER
    all(.body[IS ast::Expr].value IS ast::Constant)
LIMIT 10
```
It does yield totally irrelevant results that don't satisfy the conditions (as a proof, it yields 6x more results than the former query, which in theory should be its superset so this number should be lower/equal to the amount of items in the former set, not more.)

**Comment by isidentical** [comment]
I'm not sure but this might be related with how inline `IS` operator works, maybe? 
```

SELECT (
        FOR _item
        IN {(
            SELECT
                ast::FunctionDef {body}
            FILTER
                count(.body) = 5
            LIMIT 1
        ).body}
        UNION (
            _item IS ast::Return
        )
);
```
This query yields a set of 5 items but on the other hand this one
```
SELECT (
        FOR _item
        IN {(
            SELECT
                ast::FunctionDef {body}
            FILTER
                count(.body) = 5
            LIMIT 1
        ).body}
        UNION (
            _item[IS ast::Return].value IS ast::Name
        )
);
```
yields a set of 1 item. The weird thing is that, this is also same when we `AND` the type check as a separate check;
```
SELECT (
        FOR _item
        IN {(
            SELECT
                ast::FunctionDef {body}
            FILTER
                count(.body) = 5
            LIMIT 1
        ).body}
        UNION (
            _item IS ast::Return AND
            _item[IS ast::Return].value IS ast::Name
        )
);
```

The documentation says inline type checking removes all the items that doesn't qualify with the condition, but in this case I'd expected them to become `false` without even dropping to the inline checking part.

**Comment by elprans** [comment]
> It does yield totally irrelevant results that don't satisfy the conditions

This is probably because `all({}) == true`, and so you also need to add an `exists()` check.

**Comment by isidentical** [comment]
```
SELECT
    ast::FunctionDef
FILTER
    (
        WITH
            items := (
                FOR _item
                IN {.body}
                UNION (
                    _item IS ast::Expr
                    AND _item[IS ast::Expr].value IS ast::Constant
                )
            )
        SELECT
            (EXISTS items) AND all(items)
    )
```
I tried this query (with both this way and with only `_item[IS ast::Expr].value IS ast::Constant` instead of `_item IS ast::Expr` + this) and I still get the irrelevant results.

**Reply by isidentical** [reply]
This is a working alternative @elprans, but far from being decent. 
```
SELECT count((
    SELECT
        ast::FunctionDef
    FILTER
        (
            WITH
                items := (
                    FOR _item
                    IN {.body}
                    UNION (
                        (_item[IS ast::Return].value IS ast::Tuple ?? false)
                    )
                )
            SELECT
                all(items)
        )
));
```

**Reply by elprans** [reply]
Hm. This might be a bug.  Do you mind posting an issue with a minimal repro?
